File: codemap-intellij/src/main/kotlin/com/codemap/intellij/engine/CoreEngineBridge.kt
package com.codemap.intellij.engine

import com.codemap.intellij.model.AnalysisResult
import com.google.gson.Gson
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.project.Project
import java.io.BufferedReader
import java.io.File
import java.io.InputStreamReader

/**
 * Bridge to the CodeMap core Java engine.
 * Spawns the engine as a subprocess and captures JSON output.
 */
class CoreEngineBridge(private val project: Project) {

    private val log = Logger.getInstance(CoreEngineBridge::class.java)
    private val gson = Gson()

    /**
     * Execute a core engine command.
     */
    fun execute(command: String, target: String? = null, depth: Int = 5): AnalysisResult {
        val jarPath = resolveJarPath()
        val javaPath = resolveJavaPath()
        val projectPath = resolveSourceRoot()

        val args = mutableListOf(javaPath, "-jar", jarPath, "--project", projectPath, "--command", command)
        if (target != null) {
            args.addAll(listOf("--target", target))
        }
        args.addAll(listOf("--depth", depth.toString()))

        log.info("Executing: ${args.joinToString(" ")}")

        val process = ProcessBuilder(args)
            .directory(File(projectPath))
            .redirectErrorStream(false)
            .start()

        val stdout = BufferedReader(InputStreamReader(process.inputStream)).readText()
        val stderr = BufferedReader(InputStreamReader(process.errorStream)).readText()
        val exitCode = process.waitFor()

        if (exitCode != 0) {
            log.error("Engine failed (code $exitCode): $stderr")
            throw RuntimeException("CodeMap analysis failed: $stderr")
        }

        if (stderr.isNotBlank()) {
            log.info("Engine logs: $stderr")
        }

        return gson.fromJson(stdout, AnalysisResult::class.java)
    }

    fun getCallGraph(target: String, depth: Int = 5): AnalysisResult =
        execute("callgraph", target, depth)

    fun getIncomingCalls(target: String): AnalysisResult =
        execute("incoming-calls", target)

    fun getClassDependencies(target: String): AnalysisResult =
        execute("dependencies", target)

    fun getImpactAnalysis(target: String): AnalysisResult =
        execute("impact", target)

    fun detectCircularDependencies(): AnalysisResult =
        execute("circular-deps")

    private fun resolveJarPath(): String {
        // Look for the JAR in the project directory or default location
        val projectBase = project.basePath ?: "."
        val candidates = listOf(
            "$projectBase/codemap-core/target/codemap-core-1.0.0-SNAPSHOT.jar",
            "$projectBase/../codemap-core/target/codemap-core-1.0.0-SNAPSHOT.jar",
            System.getProperty("user.home") + "/.codemap/codemap-core.jar"
        )
        return candidates.firstOrNull { File(it).exists() }
            ?: throw RuntimeException("codemap-core JAR not found. Build the core engine with 'mvn package'.")
    }

    private fun resolveJavaPath(): String {
        val javaHome = System.getProperty("java.home") ?: System.getenv("JAVA_HOME") ?: ""
        return if (javaHome.isNotEmpty()) "$javaHome/bin/java" else "java"
    }

    private fun resolveSourceRoot(): String {
        val basePath = project.basePath ?: throw RuntimeException("No project path available")
        // Try common source roots
        val candidates = listOf(
            "$basePath/src/main/java",
            "$basePath/src",
            basePath
        )
        return candidates.firstOrNull { File(it).exists() } ?: basePath
    }
}

File: codemap-vscode/src/engine/coreEngineBridge.ts
import * as vscode from 'vscode';
import * as cp from 'child_process';
import * as path from 'path';
import { AnalysisResult } from '../types/graph';

/**
 * Bridge to the CodeMap core Java engine.
 * Invokes the engine via CLI (java -jar) and parses JSON output.
 */
export class CoreEngineBridge {
    private jarPath: string;
    private javaPath: string;

    constructor() {
        const config = vscode.workspace.getConfiguration('codemap');
        this.jarPath = config.get<string>('coreJarPath') || this.resolveDefaultJarPath();
        this.javaPath = this.resolveJavaPath(config.get<string>('javaHome') || '');
    }

    /**
     * Execute a core engine command and return the parsed result.
     */
    async execute(command: string, projectPath: string, target?: string, depth?: number): Promise<AnalysisResult> {
        const args = [
            '-jar', this.jarPath,
            '--project', projectPath,
            '--command', command
        ];

        if (target) {
            args.push('--target', target);
        }
        if (depth !== undefined) {
            args.push('--depth', depth.toString());
        }

        return new Promise((resolve, reject) => {
            const process = cp.spawn(this.javaPath, args, {
                cwd: projectPath,
                env: { ...global.process.env }
            });

            let stdout = '';
            let stderr = '';

            process.stdout.on('data', (data: Buffer) => {
                stdout += data.toString();
            });

            process.stderr.on('data', (data: Buffer) => {
                stderr += data.toString();
            });

            process.on('close', (code: number | null) => {
                if (code === 0) {
                    try {
                        const result: AnalysisResult = JSON.parse(stdout);
                        resolve(result);
                    } catch (parseError) {
                        reject(new Error(`Failed to parse engine output: ${parseError}`));
                    }
                } else {
                    reject(new Error(`Engine exited with code ${code}: ${stderr}`));
                }
            });

            process.on('error', (err: Error) => {
                reject(new Error(`Failed to start engine: ${err.message}. Ensure Java is installed and codemap-core JAR path is configured.`));
            });
        });
    }

    async getCallGraph(projectPath: string, methodSignature: string, depth: number): Promise<AnalysisResult> {
        return this.execute('callgraph', projectPath, methodSignature, depth);
    }

    async getIncomingCalls(projectPath: string, methodSignature: string): Promise<AnalysisResult> {
        return this.execute('incoming-calls', projectPath, methodSignature);
    }

    async getClassDependencies(projectPath: string, className: string): Promise<AnalysisResult> {
        return this.execute('dependencies', projectPath, className);
    }

    async getImpactAnalysis(projectPath: string, className: string): Promise<AnalysisResult> {
        return this.execute('impact', projectPath, className);
    }

    private resolveDefaultJarPath(): string {
        // Look for JAR relative to extension location
        return path.join(__dirname, '..', '..', 'codemap-core', 'target', 'codemap-core-1.0.0-SNAPSHOT.jar');
    }

    private resolveJavaPath(javaHome: string): string {
        if (javaHome) {
            return path.join(javaHome, 'bin', 'java');
        }
        return process.env.JAVA_HOME
            ? path.join(process.env.JAVA_HOME, 'bin', 'java')
            : 'java';
    }
}
